db/range_del_aggregator.cc

```diff
@@ -328,15 +342,29 @@ class CollapsedRangeDelMap : public RangeDelMap {
     }
 
     if (t.seq_ == prev_seq()) {
-      // The new tombstone is unterminated in the map:
-      //
-      //     3:             OR   3: --G       OR   3: --G   K--
-      //     2: C-------k        2:   G---k        2:   G---k
-      //                  ^                 ^               ^
-      // End it now, returning to the last seqno we covered. Because end keys
-      // are exclusive, if there's an existing transition at t.end_key_, it
-      // takes precedence over the transition that we install here.
-      rep_.emplace(t.end_key_, end_seq);  // emplace is a noop if existing entry
+      // The new tombstone is unterminated in the map.
+      if (it != rep_.end() && t.seq_ == it->second && ucmp_->Compare(it->first, t.end_key_) == 0) {
+        // The new tombstone ends at the start of another tombstone with an
+        // identical seqno. Merge the tombstones by removing the existing
+        // tombstone's start key.
+        rep_.erase(it);
+      } else if ((it == rep_.end() || end_seq != it->second) && end_seq != prev_seq()) {
```

Comment:
nit: would it be a bit more clear to switch the positions of the `else if` and `else` clauses? I think the negation of the else if condition with a bit of equivalent modification to `end_seq == t.seq_ || (it != rep_.end() && end_seq == it->second)` makes it a bit more clear that we have overlapping tombstones with the same seqnum, while I had to stare at this version for a while before convincing myself that it was valid to emplace at the end key.

Reply:
Hmm, actually, when would it be valid for `end_seq == it->second`? If I understand all this correctly, before the new tombstone was added to the map, the entry with `end_seq` was immediately before `it`, which implies that their seqnums should be different.

Reply:
> Hmm, actually, when would it be valid for end_seq == it->second? If I understand all this correctly, before the new tombstone was added to the map, the entry with end_seq was immediately before it, which implies that their seqnums should be different.

This occurs when you're adding a tombstone like [b, c) @ 5 when a tombstone like [a, d) @ 5 already exists in the map. In this case we want to avoid adding the end point for a map of
```
a -> 5
d -> 0
```

instead of:

```
a -> 5
c -> 0
d -> 0
```

Also, pushed a new commit that reverses the conditional!

Reply:
Ah, got it. Thanks for the explanation!